appinventor/appengine/src/com/google/appinventor/client/editor/ProjectEditor.java

```diff
@@ -307,28 +309,73 @@ public final void changeProjectSettingsProperty(String category, String name, St
    */
 
   public final void recordLocationSetting(String componentName, String newValue) {
-    OdeLog.log("ProjectEditor: recordLocationSetting(" + componentName + "," + newValue + ")");
-    locationHashMap.put(componentName, newValue);
-    recomputeLocationPermission();
+    recordPermission(componentName, newValue, SettingsConstants.YOUNG_ANDROID_SETTINGS_USES_LOCATION, locationHashMap);
   }
 
-  private final void recomputeLocationPermission() {
-    String usesLocation = "False";
-    for (String c : locationHashMap.values()) {
-      OdeLog.log("ProjectEditor:recomputeLocationPermission: " + c);
+  public final void recordCameraSetting(String componentName, String newValue) {
+    recordPermission(componentName, newValue, SettingsConstants.YOUNG_ANDROID_SETTINGS_USES_CAMERA, cameraHashMap);
+  }
+
+  public final void recordMicrophoneSetting(String componentName, String newValue) {
+    recordPermission(componentName, newValue, SettingsConstants.YOUNG_ANDROID_SETTINGS_USES_MICROPHONE, microphoneHashMap);
+  }
+
+  /**
+   * Auxiliary function to record a targetted permission for a component.
+   *
+   * @param componentName Name of the component to record permission for
+   * @param newValue New value of the permission
+   * @param permission Name of the permission
+   * @param permissionMap Map holding the permission mappings (component -> value)
+   */
+  private final void recordPermission(String componentName, String newValue, String permission,
+                                 Map<String, String> permissionMap) {
+      OdeLog.log("ProjectEditor: record" + permission + "Setting(" + componentName + "," + newValue + ")");
```

Comment:
Indentation should be 2 space from function declaration (4 spaces total)

Reply:
I think we should remove this call to OdeLog.log or replace it with Ode.CLog. OdeLog manipulates the DOM and so every log message adds HTML, making the UI slower.